### ISSUE TITLE
[ENH]: should Axes hold references to widgets?

### ISSUE THREAD

**Issue by timhoffm** [matplotlib/matplotlib#24373]
### Problem

People often forget to hold a reference to a widget so that it‘s getting garbage collected, c.f. #24373.

### Proposed solution

Track widgets in a (maybe private) Axes attribute, as we do for artists?

The widget‘s artists (=visual part of the widget) are referenced, so there is a tight relationship. We should add the widget objects themselves (=logic part of the widget) as well.

**Comment by anntzer** [comment]
See also https://github.com/matplotlib/matplotlib/pull/16221 and linked issues.

**Comment by timhoffm** [comment]
Thanks for the reminder on #16221. At least for the case of widgets, I feel that holding explicit references is simpler than the lambda construct (explicit is better than implicit).

**Comment by anntzer** [comment]
To be clear, I don't have a strong preference implementation-wise, it was just to point out an earlier discussion on the same topic.

**Comment by efiring** [comment]
@anntzer in #16221 you complain about the use of weak references in the callback registry. Proposals to address the current issue include your lambda trick and tracking strong references in the Figure or Axes. Do you also have in mind a solution based on switching from weak to strong references in the callback registry?  Or would that have been your preference from the start, but it is now too late to change it?

**Comment by anntzer** [comment]
I would certainly prefer using strong references (the lambda trick is basically just a way to get strong references for cheap).  I think it's possible to even have a proper deprecation period when changing that (basically emit a warning if a method proxy gets gc'ed from the callbackregistry if it didn't get disconnected before -- untested, though).
However IIRC @tacaswell is strongly (hehe) in favor of keeping the weakref semantics.

**Comment by carlosgmartin** [comment]
I agree with @anntzer's comments [here](https://github.com/matplotlib/matplotlib/pull/16221#issuecomment-575565228). The current behavior is surprising and not user-friendly.

**Comment by anntzer** [comment]
See also #26881 ("Provide a standard place to manually register widgets/animations on a figure/artist/etc.") for an alternative suggestion.
